### DISCUSSION TITLE
2025 hackathon projects

### DISCUSSION THREAD

**Discussion by tachella** [deepinv/deepinv#700] [Ideas]
## Hackathon projects

Some projects for the hackathon that will take place on September 8-10 in Marseille, France.

**Comment by mh-nguyen712** [comment]
A possible project

### Project 4: benchmarking and improving the performance of the library.

**Idea**: Benchmark and improve (if possible) some core functionalities of the library. 

As the library becomes more and more complex with new solvers/functionalities, it would be great if the core of the library has the best performance (in term of compute time), e.g. physics classes, some adjoint operators, some autodiff functions, ...


**Help**: We can start by creating a utility function that measures the compute time/peak memory usage/... based on torch (eg torch. Timer).
Then if possible, use [pytorch profiler](https://docs.pytorch.org/tutorials/intermediate/tensorboard_profiler_tutorial.html) to profile, eg an unfolder solver.  

**Technical challenges**: function-by-function profiling, time consuming and diffucult task to improve the performance. 

**Impact**: This makes the library faster! 

**Steps**:
1. Write utility functions (eg in deepinv.utils.perf) for time/memory quick report
2. Profile some core functions (e.g. Trainer)
3. Detect the botteneck and improve.

**Closes**: #601

**Comment by tachella** [comment]
### Project 1: Multiview physics

**Idea**: Implement a new physics operator that combines measurements from different 'views' of the same object.

Define a linear multiview operator which measures $J$ measurement vectors $y_1,\dots,y_J$ of a single image $x$, following the model

$y_j = AT_jx \text{ for } j=1,\dots,J$

where $A$ is an arbitrary forward operator (eg downsampling), and $T_j$ are affine transformations defined by $2\times 3$ matrices, which include translations, rotations and shearing.


**Help**: Use torch's [grid_sample](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html) function to implement the transformations $T_j$.
You can also have a look at a preliminary implementation as part of a tutorial [here](https://colab.research.google.com/drive/1mK9Ame4G1-4d40GnP7wpuqPlAFDw4rzd).

**Technical challenges**: If time allows, it would be nice to write some basic code that estimates the transformations $T_j$ from the measurements $y_j$.

**Impact**: This operator would be useful for multiview imaging problems such as multiframe super-resolution, where we have multiple measurements of the same object from different viewpoints.

**Steps**:
1. Write the physics operator, and make sure it works for a simple problem
2. Write some pytests checking with different base operators $A$.
3. Write a sphinx gallery demo showing how to use the operator for multi-frame super-resolution.

**Comment by tachella** [comment]
### Project 2: Custom autodiff for linear solvers

**Idea**: Implement a custom `backward` function for the [least_squares](https://deepinv.github.io/deepinv/api/stubs/deepinv.optim.utils.least_squares.html#deepinv.optim.utils.least_squares) function

Seeing the linear solver as a function 

$f(y, z, \theta, \gamma) = \text{arg}\min_x \|A_{\theta}x-y\|^2 + \frac{1}{\gamma}\|x-z\|^2$
 
We would like to compute the closed-form gradients with respect to inputs $(y, z, \theta, \gamma)$
by reusing the linear solver itself, and avoiding storing 
intermediate variables during the forward pass (what is currently done automatically by PyTorch).

**Impact**: This option would be useful for unrolled reconstruction methods where we require
evaluating a proximal operator (which is often a least-squares problem) in the forward pass, and 
backpropagating through it in the backward pass. This is currently done by storing all intermediate variables during the forward pass, which can be very memory-intensive.

**Technical challenges**: Matrix $A_{\theta}$ is passed as a callback and may depend on parameters $\theta$ 

**Steps**:
1. Understand the closed-form gradients of $f$ w.r.t. $(y, z, \theta, \gamma)$.
1. Write a custom `torch.nn.Function` replacing the backward of `least_squares` by the closed-form one.
2. Write tests to check that gradients are calculated properly.

**Reply by mh-nguyen712** [reply]
I'm working on this one if anyone want to join me!

**Reply by mh-nguyen712** [reply]
This is done and merged in #739

**Comment by tachella** [comment]
### Project 3: Latent space models

**Idea**: It is becoming increasingly popular to solve inverse problems in
the latent space of some VQ-GAN autoencoder $g$, that is solving the non-linear inverse problem

$y = A  g(z)$

where $A$ is the (generally) linear physics and $z$ is the latent code, such that
$\hat{x}=g(z)$ is the reconstructed image in pixel space.
This problem can be solved via any of the methods in the library, including diffusion, PnP, unrolling, etc.

**Technical challenges**: Find a modular and backward-compatible way of integrating this new functionality into
the library, without having to rewrite most reconstruction methods.

**Impact**: This would enable state-of-the-art reconstruction methods based on latent space models.

**Steps**
1. Find a SOTA implementation of a VQ-GAN autoencoder, and adapt a minimal version of it to work with deepinv.
2. Explore different ways of integrating this into the library, either by creating new non-linear composed physics $A \circ g$, or by modifying existing methods to work with latent space models.
3. Write a sphinx gallery demo showing how to use the new functionality.
4. Write tests to check that everything works as expected.

**Comment by bemc22** [comment]
### Project 5: Spatial Unwrapping

**Idea:** Include a new physics model for spatial unwrapping. Spatial unwrapping is a non-linear inverse problem of great interest in different communities (phase unwrapping from optical interferometry, MRI, fringe pattern profilometry, InSAR, and recently, modulo imaging). This is modeled with the form:

$$\mathbf{y} = M(x,k) = \mathbf{x} - k \cdot q(\mathbf{x}/k)$$

where $k$ is the folding factor and $q(\cdot)$ is the quotient function, which could be the round or floor operators, among others. In practice, since the modulo is a non-linear operator, this problem is solved with linear optimization algorithms under the assumption of the Itoh condition by minimizing the following measurement consistency term through application of the finite discrete difference operator $\Delta$:

$$\min_{\mathbf{x}} \Vert  M( \Delta \mathbf{y},k) -  \Delta \mathbf{x} \Vert_2^2$$

For this case, it is interesting to analyze that the forward model and the fidelity terms have different operators!

**Technical Challenge:** Propose and implement `DataFidelity` and `Physics` subclasses for the specific requirements of spatial unwrapping that are compatible with current restoration algorithms in the library.

**Impact:** This would enable different spatial unwrapping problems to have access to state-of-the-art restoration algorithms in image processing.

**Steps:**

1. Implement the `Physics` for spatial unwrapping with variable quotient function (round or floor)
2. Propose and discuss the best way to include Itoh Fidelity for spatial unwrapping (A custom `DataFidelity`?)
3. Validate compatibility with restoration algorithms such as PGD, FISTA, PnP, ADMM, etc.
4. Write demos for specific inverse problems such as phase unwrapping and modulo imaging
5. Write tests to check overall integrity

**Comment by nducros** [comment]
### Project 6: Connect SPyRiT to deepinv

**Idea**:

Enable the use of the [SPyRiT](https://spyrit.readthedocs.io/en/3.0.2/) package together with deepinv.

SPyRiT is an open-source PyTorch-based package for single-pixel imaging. This integration would allow users to leverage the capabilities of SPyRiT (e.g., forward models corresponding to actual imaging configurations, access to experimental datasets) while utilizing the optimization algorithms available in deepinv.

**Technical Challenge**:

Propose and implement Physics subclasses ?

**Impact**:

This would allow people in the optics community to access a wider variety of state-of-the-art reconstruction algorithms. Conversely, it would allow people in the inverse problems community to easily access more realistic forward models and the corresponding experimental datasets.

**Steps**:

1. Discuss the best way to call SPyRiT forward models from deepinv (a custom Physics subclass?).
2. Implement the proposed solution with a simple reconstruction algorithm.
3. Validate the implementation using a simple reconstruction algorithm (to be determined) available in both SPyRiT and deepinv. Note: SPyRiT also includes several reconstruction algorithms (see this [paper](https://doi.org/10.1364/OE.559227)).
4. Add a tutorial to both the deepinv and SPyRiT documentation.

**Reply by Andrewwango** [reply]
Note: the example should be added to the external-libraries folder.

**Reply by tbaudier** [reply]
Hello,
With @nducros we just added an example in spyrit with the connection to deepinv. It's a python script with a basic reconstruction workflow:
https://github.com/openspyrit/spyrit/blob/master/tutorial/wip/tuto_a00_connect_deepinv.py
That script is used to generate that tutorial in the doc:
https://spyrit.readthedocs.io/en/master/gallery/tuto_a00_connect_deepinv.html
And also a test in the CI:
https://github.com/openspyrit/spyrit/actions/runs/17949510097/job/51044569698
In that test we do not check the output, it's a test to know if the compatibility between spyrit and deepinv is still ok.

Do you want me to upload that test in your external-libraries folder?

**Comment by Andrewwango** [comment]
### Project 7: Linting

**Idea**:

DeepInverse is written in high quality Python, and avoids common pitfalls and bugs (such as https://github.com/deepinv/deepinv/issues/628). 

**Steps**:

1. We should extend our usage of `ruff` to both its [default set](https://github.com/astral-sh/ruff#:~:text=select%20%3D%20%5B%22E4%22%2C%20%22E7%22%2C%20%22E9%22%2C%20%22F%22%5D) and some more rules taken from [Python anti-patterns](https://docs.quantifiedcode.com/python-anti-patterns/index.html): `SLF001
, BLE001
, N805
, PLE0101
, PLR6301
, PLW0120
, F404
, PLW0603
, S102
, E721`

2. We can add more rules to cover the acceptance criteria of this issue: https://github.com/deepinv/deepinv/issues/539

3. Finally, we should add these rules to our workflow and check that DeepInverse passes them all.

Further notes: [ruff-pylint correspondance](https://github.com/astral-sh/ruff/issues/970)

**Reply by Andrewwango** [reply]
See #705 for progress!

**Reply by MaximeSongIdris** [reply]
default set = ["E4", "E7", "E9", "F"]

**Comment by Andrewwango** [comment]
### Project 8: External integrations

**Idea**:

This project has two streams.

Stream 1: interfaces for external libs:

- Physics ([MRI-NUFFT](https://mind-inria.github.io/mri-nufft/index.html), Astra (done))
- Data loading (napari, ISMRMRD/H5 standard, DICOM (see [dosma](https://dosma.readthedocs.io/en/latest/generated/dosma.MedicalVolume.html#dosma.MedicalVolume) ([tutorial](https://colab.research.google.com/drive/1zY5-3ZyTBrn7hoGE5lH0IoQqBzumzP1i?usp=sharing#scrollTo=Z6oeR_VEVbFL))), NIFTI (nibabel), remote sensing (large geotiff streaming with `rasterio`/`rioxarray` ([sen2naip](https://huggingface.co/datasets/isp-uv-es/SEN2NAIP) dataset); cosar with [this gist](https://gist.github.com/Andrewwango/7c47cc62fb8a1a5f1770833c2c7984c2)). To be placed into `deepinv.utils.io` and maybe linked to `deepinv.io`

Stream 2: DeepInverse plugins for existing industry software: e.g. ImageJ (microscopy)

**Impact**:

Stream 1: we open up DeepInverse to practitioners/researchers in more industries by making it more seamless to load and inference industry-specific images

Stream 2: we spread awareness of DeepInverse by providing easy-to-use deep learning functionality for users of existing imaging software in various industries.

**Reply by chaithyagr** [reply]
I would like to contribute for bindings with mri-nufft.

**Reply by nucli-vicky** [reply]
I can add nibabel

**Reply by nucli-vicky** [reply]
Project 16 is partially related to this. Requires integrating STIR / Interfile IO

**Reply by Andrewwango** [reply]
Progress: see #768

**Comment by Andrewwango** [comment]
### Project 9: Refactor Trainer

**Idea**:

The Trainer code is expanding and the current implementation might not be the most optimal in terms of:

- Maintainability (bug awareness, test combination explosion)
- Scalable (as more features are added) while remaining modular
- Flexible/extendable for users to adapt the trainer to their needs
- Readable
 
Context for existing discussions and ideas (@jscanvic):

- https://github.com/deepinv/deepinv/discussions/411
- https://github.com/deepinv/deepinv/issues/545
- https://github.com/deepinv/deepinv/issues/553

**Technical Challenge**:

Tackling this problem will require some high-level Python design thinking, not just simple scientific coding. Potential solutions:

- Hooks, callbacks of configs?
- Unite with/inherit from PyTorch Lightning/ Ignite trainer?
- Non-OOP-oriented solutions in accordance with advanced Python design principles, such as decorators or adaptors: see [The Composition Over Inheritance Principle](https://python-patterns.guide/gang-of-four/composition-over-inheritance/)

**Reply by Tmodrzyk** [reply]
Here's the link to brainstorming notes [https://docs.google.com/document/d/12fZKxPQKAYreWcBQT6pHV2ojfIqtu0b3sphNdBNJKNQ/edit?usp=sharing](https://docs.google.com/document/d/12fZKxPQKAYreWcBQT6pHV2ojfIqtu0b3sphNdBNJKNQ/edit?usp=sharing)

**Comment by samuro95** [comment]
### Project 10: PnP priors with general restoration models. 

#### Idea:

Building on the Plug-and-Play module, the idea is to use other restoration operators in order to solve inverse problems.

#### Reference papers:
- [A Restoration Network as an Implicit Prior](https://arxiv.org/pdf/2310.01391)
- [FiRe: Fixed-points of Restoration Priors for Solving Inverse Problems](https://openaccess.thecvf.com/content/CVPR2025/html/Terris_FiRe_Fixed-points_of_Restoration_Priors_for_Solving_Inverse_Problems_CVPR_2025_paper.html)

####  Plan of Work  
1. **Model Selection**  
   Start with an open-weight restoration network trained for a corruption other than denoising. A possible candidate is the *RAM* general restoration model, already available in the library.  

2. **PnP Integration**  
   Implement an example where this restoration model is used within the existing PnP prior module and in a standard optimization algorithm, already implemented in the library. 

3. **Parameter Tuning**  
   Optimize hyperparameters to achieve good performance on a selected inverse problem of your choice.  

4. **Algorithmic Exploration**  
   Extend the study by testing the approach within alternative optimization algorithms, possibly in collaboration with the team working on stochastic Plug-and-Play methods.

**Comment by samuro95** [comment]
### Project 11: Stochastic PnP methods.

#### Idea: implement PnP methods that add noise or apply more general random transformations before denoising / restoration.

#### Reference papers:
- [Plug-and-Play image restoration with Stochastic deNOising REgularization](https://arxiv.org/pdf/2402.01779)
- [FiRe: Fixed-points of Restoration Priors for Solving Inverse Problems](https://openaccess.thecvf.com/content/CVPR2025/html/Terris_FiRe_Fixed-points_of_Restoration_Priors_for_Solving_Inverse_Problems_CVPR_2025_paper.html)

The goal of this project is to implement a method that add stochasticity inside PnP algorithms. The team can for example choose between SNORE (add nose before denoising) or FIRE (apply more general transformations) and will work in collaboration with the team of Project 10. 

The team can choose to either integrate general stochastic updates inside the DeepInv Optim module or to directly build an example file for the method of their choice.

**Comment by matthieutrs** [comment]
### Project 12 - SotA superresolution

#### Idea:
SR methods in the library are suboptimal today. I think doing some re-evaluations / benchmarks of the library for SR would be of interest, as well as adding latest SOTA methods. Below are some SOTA models that are currently used in the literature, by order of perf/age:

### Plan of work:
1. Implement [HYPIR](https://github.com/XPixelGroup/HYPIR)
2. (Optional) Implement [DiffBIR](https://github.com/XPixelGroup/DiffBIR)
3. (Optional) Implement [StableSR](https://github.com/IceClear/StableSR)
4. (Optional) Implement [ESRGAN](https://arxiv.org/abs/1809.00219)/SRGAN - also showcases popular GANs. We already have the GAN framework in deepinverse.

### Technicality

If the implementation of deepinv is matching perfectly that of the diffusers library, things should be rather straightforward. However, things are likely going to be bumpy 🙃

**Comment by alexspagn** [comment]
### Project 13 - PnP Latent Diffusion Inverse Solvers

**Idea:**

Modern PnP SOTA methods for image restoration exploit Latent Diffusion Models (LDMs) and distilled versions of them (e.g., Consistency Models) to inject prior information during the restoration process. The idea is to implement these methods ([LDPS/PSLD](https://arxiv.org/abs/2307.00619), [P2L](https://arxiv.org/abs/2310.01110), [TReg](https://arxiv.org/abs/2311.15658), [LATINO-PRO](https://arxiv.org/abs/2503.12615)) under the same framework, mimicking what has already been implemented with the DPS algorithm. The diffusers library can provide the implementation of the VAEs and score-networks for the priors ([Stable Diffusion v1.5](https://arxiv.org/abs/2112.10752), [Stable Diffusion XL](https://arxiv.org/abs/2307.01952), and [DMD2](https://arxiv.org/abs/2405.14867)).

All these methods are already implemented in [this](https://github.com/LATINO-PRO/LATINO-PRO) repo; the key point would be to integrate them in deepinv.

### Plan of work

1. Think about the structure of classes we will need to implement the methods
2. Check what's already available in deepinv (e.g. the Reconstructor class) and what we can evolve from this, identify new classes to implement
3. Specify the tests we want to run (Deblur, SR, Inpainting...) and test all the models
5. (If time allows) Create examples/tutorials (e.g. a notebook that runs multiple methods and compares the results)

**Impact:**

Allow researchers/practitioners to easily run SOTA restoration methods all in the same environment, enabling customizations to specific uses. Some of these methods have never been implemented with high-resolution priors (1024x1024 res.) on publicly available repositories before, and I believe that the community would appreciate this effort.

**Comment by CoConstan** [comment]
Project 14: Connect Pytomography to deepinv
Idea:

Enable the use of the [Pytomography](https://pytomography.readthedocs.io/en/latest/) package together with deepinv.

Pytomography is an open-source PyTorch-based package for mainly SPECT/CT reconstruction. This integration would allow users to leverage the capabilities of Pytomography (e.g. exact ECT forward models corresponding to actual imaging configurations) while utilizing the optimization algorithms available in deepinv.

Technical Challenge:

Propose and implement Physics subclasses ?
3D problem => high VRAM usage

Impact:

This would allow people in the SPECT community to access a wider variety of state-of-the-art reconstruction algorithms. Conversely, it would allow people in the inverse problems community to easily access more realistic forward models for 3D problems.

**Comment by chaithyagr** [comment]
Project 15:[Application] MRI reconstruction for accelerated non-Cartesian imaging

- MRI-NUFFT bindings
- PnP with HQS model w/ and wo/ pre-conditioners (https://arxiv.org/abs/2411.01955), example
- Wavelet / TV based CS reconstructions (FISTA)
- EBM: Try to bring some EBM models up .. towards a comprehensive benchmark on prospective data.

**Comment by nucli-vicky** [comment]
Project 16: Add PET data and physics model

Idea: 
Add the PETRIC dataset to DeepInverse. Contains sinograms from a variety of scanners. Will require a new physics model which takes into account the additive term (scatter & randoms). Will also require interfacing with the specific file format (or converting to a more general file format).

Plan of work:
1. Create an interface to work with the PETRIC's file format
2. Test models using projection operator only
3. Define the new physics

**Reply by nucli-vicky** [reply]
have to integrate [parallelproj](https://github.com/gschramm/parallelproj)
